zebra-consensus/src/chain.rs

```diff
@@ -72,34 +70,38 @@ where
     type Future =
         Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>>;
 
-    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
         // Correctness:
         //
-        // We can't call `poll_ready` on the block and checkpoint verifiers here,
-        // because each `poll_ready` must be followed by a `call`, and we don't
-        // know which verifier we're going to choose yet.
-        // See #1593 for details.
+        // We acquire checkpoint readiness before block readiness, to avoid an unlikely
+        // hang during the checkpoint to block verifier transition. If the checkpoint and
+        // block verifiers are contending for the same buffer/batch, we want the checkpoint
+        // verifier to win, so that checkpoint verification completes, and block verification
+        // can start. (Buffers and batches have multiple slots, so this contention is unlikely.)
+        use futures::ready;
+        // The chain verifier holds one slot in each verifier, for each concurrent task.
+        // Therefore, any shared buffers or batches polled by these verifiers should double
+        // their bounds. (For example, the state service buffer.)
+        ready!(self
+            .checkpoint
+            .poll_ready(cx)
+            .map_err(VerifyChainError::Checkpoint))?;
+        ready!(self.block.poll_ready(cx).map_err(VerifyChainError::Block))?;
```

Comment:
This code is new, it is designed to satisfy the latest constraints in #1593:

1. `poll_ready` must be called **at least once** for each `call` 
2. Once we've reserved a buffer slot, we always get `Poll::Ready` from a buffer, regardless of the current readiness of the buffer or its underlying service
3. The `Buffer`/`Batch` capacity limits the number of concurrently waiting tasks. Once this limit is reached, further tasks will block, awaiting a free reservation.
4. Some tasks can depend on other tasks before they resolve. (For example: block validation.) If there are task dependencies, **the `Buffer`/`Batch` capacity must be larger than the maximum number of concurrently waiting tasks**, or Zebra could deadlock (hang).
5. **To avoid deadlocks, we should acquire slots in a consistent order, using the `ready!` macro**, rather than polling them all simultaneously